### PR TITLE
terratest abstraction

### PR DESCRIPTION
**Pull request by kuritka**
I created test absatraction for terratests. The change is backward compatible so terratests can be still written an old fashioned way.

I had to create package `terratest/utils` containing
- test abstraction
- helper functions
- TestSettings struct

Also I had to move go.mod file from terratest/test to terratest.

I removed basic failover and basic roundrobin tests and replaced by full roundrobin and full failover tests. See the table below:

|                                           | roundrobin<br>**k3d-test-gslb1 (EU)** | roundrobin<br>**k3d-test-gslb2 (US)** | failover<br>**k3d-test-gslb1 (EU)** | failover<br>**k3d-test-gslb2 (US)** |
|-------------------------------------------|:-------------------------------------:|:-------------------------------------:|:-----------------------------------:|:-----------------------------------:|
| IP of CoreDNS                             |             `10.43.18.112`            |             `10.43.56.78`             |            `10.43.18.112`           |            `10.43.56.78`            |
| Localtargets                              |            `10.43.101.181`            |            `10.43.148.142`            |           `10.43.101.181`           |           `10.43.148.142`           |
| initial state                             |   `10.43.101.181`<br>`10.43.148.142`  |   `10.43.101.181`<br>`10.43.148.142`  |           `10.43.101.181`           |           `10.43.101.181`           |
| k3d-test-gslb2, <br>`make stop-test-app`  |            `10.43.148.142`            |            `10.43.148.142`            |           `10.43.101.181`           |           `10.43.101.181`           |
| k3d-test-gslb1, <br>`make stop-test-app`  |                  `-`                  |                  `-`                  |                 `-`                 |                 `-`                 |
| k3d-test-gslb2, <br>`make start-test-app` |            `10.43.101.181`            |            `10.43.101.181`            |           `10.43.148.142`           |           `10.43.148.142`           |
| k3d-test-gslb1, <br>`make start-test-app` |   `10.43.101.181`<br>`10.43.148.142`  |   `10.43.148.142`<br>`10.43.101.181`  |           `10.43.101.181`           |           `10.43.101.181`           |

**The IP's addresses may differ*

Because the tests have more steps than before, terratest takes more time now. In the future, however, we can optimize this by running individual steps (e.g : create instance) on both clusters in parallel. This will reduce time.
Signed-off-by: kuritka <kuritka@gmail.com>

### CODE REVIEW

File: terratest/utils/extensions.go [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+func waitForLocalGSLBNew(t *testing.T, host string, port int, expectedResult []string) (output []string, err error) {
+	return DoWithRetryWaitingForValueE(
+		t,
+		"Wait for failover to happen and coredns to pickup new values...",
+		100,
+		time.Second*1,
+		func() ([]string, error) { return dns.Dig("localhost:"+strconv.Itoa(port), host) },
+		expectedResult)
+}
```
Comment (ytsarev) [inline terratest/utils/extensions.go:282]: @kuritka overall it looks great. Can we document the testing DLS functions in this file and/or create short documentation?

File: terratest/README.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
+# K8GB terratests
+Terratests are another layer in k8gb testing. They test the state of k8b on tuple of clusters and examine the behavior
```
Comment (ytsarev) [inline terratest/README.md:2]: `s/k8b/k8gb`

File: terratest/README.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
+# K8GB terratests
+Terratests are another layer in k8gb testing. They test the state of k8b on tuple of clusters and examine the behavior 
+for individual usecases. They replace manual tests; install GSLB in a separate namespace and check if everything
```
Comment (ytsarev) [inline terratest/README.md:3]:
```suggestion
for individual use cases. They replace manual tests, install GSLB in a separate namespace and check if everything 
```

File: terratest/README.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
+# K8GB terratests
+Terratests are another layer in k8gb testing. They test the state of k8b on tuple of clusters and examine the behavior 
+for individual usecases. They replace manual tests; install GSLB in a separate namespace and check if everything 
+behaves as expected.
+
+Terratest is basically a separate project inside k8gb. The tests are running as Gihub actions when interacting
```
Comment (ytsarev) [inline terratest/README.md:6]:
```suggestion
Terratest is basically a separate project inside k8gb. The tests are running as GitHub Actions when interacting 
```

File: terratest/README.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+- `/test/` contains the individual tests. Each test usually creates its own namespace where it installs its own 
+  version of GSLB. The test then runs against such instance. For optimization reasons, all tests are running 
+  in parallel. Keep this on mind!
+- `/examples/` contains yaml configurations (GSLB, ingresses) for individual tests.
+- `/utils/` contains a common framework that makes writing tests easier.
+
+## Terratest Common Framework
+First of all, nothing forces you to use the framework. There are certainly many cases where it is better not to use 
+the framework and instead interact with the cluster directly using terratest (which is itself a powerful framework).
+On the other hand, the ability to quickly spin-up clusters with test applications and then easily read ip addresses
```
Comment (ytsarev) [inline terratest/README.md:21]:
```suggestion
On the other hand, the ability to quickly spin up clusters with test applications and then easily read ip addresses 
```

**Review by ytsarev — CHANGES_REQUESTED** [review]
Just quickly went through the doc but then realized that it is probably not finished. Left the comments with suggestions I spotted so far

**Comment by kuritka** [comment]
@ytsarev, thanks for review. I amended the changes and also added test example and troubleshoot part.

**Review by ytsarev — APPROVED** [review]

**Pull request merged**